src/logsumexp.jl

```diff
@@ -126,3 +126,18 @@ function _logsumexp_onepass_op(xmax1::T, xmax2::T, r1::R, r2::R) where {T<:Numbe
     end
     return xmax, r
 end
+
+"""
+$(SIGNATURES)
+
+Compute `log.(sum!(r, exp.(X); dims=dims))`, summing over the singleton dimensions of `r`,
+and overwriting `r` with the results.
+The calculation is done in a numerically stable way that avoids intermediate
+over- and underflow.
+"""
+function logsumexp!(r::AbstractArray, X::AbstractArray)
+    m = maximum!(similar(r), X)
+    c = sum!(similar(r), X .== m)
+    sum!(r, (X .< m) .* exp.(X .- m))
+    return r .= log.(c) .+ log1p.(r ./ c) .+ m
```

Comment:
Can you benchmark the method? There are so many additional allocations that intuitively I assume it will be slower than the out-of-place method for a wide range of sizes of `X` and `r`.

Reply:
Yeah it is slower. I'll post some benchmarks for reference.